### PR TITLE
API type structure

### PR DESCRIPTION
**Pull request by lavalamp**
#15730 talked about client package arrangement. This is about the general package structure, usable in both server and client.

I think the following people will be interested in this: @thockin @bgrant0607 @krousey @smarterclayton @wojtek-t

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/16062)

### CODE REVIEW

File: docs/proposals/api-types-package-structure.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+
+In the future, it'd be nice if we could generate REST install code based on
+comment annotations in the go types.
+
+## Desired toolchain
+
+Along with packages for runtime use, we will supply tools to do the various
+generations.
+
+Ideally, the tool will take the root `apis/` directory, discover the `group`s
```
Comment (smarterclayton) [inline docs/proposals/api-types-package-structure.md]: Note that we need to be able to call and invoke these generations and tools from within the OpenShift codebase, and our structure today is:

```
pkg
└── resourcename1
    └── api
        └── v1
            ├── types.go
            └── register.go
└── resourcename2
    └── api
        └── v1
            ├── types.go
            └── register.go
└── api
    ├── register.go
    └── v1
        ├── register.go
        ├── deep_copy_generated.go
        └── conversions_generated.go
```

I'm fine with us switching to `pkg/apis/ourgroup/v1`, but can we ensure that our per type resource separation doesn't get broken (the generators support it today)?

File: docs/proposals/api-types-package-structure.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
+little else. In particular, it *must not* import its parent `apis/group`.
```
Comment (krousey) [inline docs/proposals/api-types-package-structure.md]: So will this be the only package needed to use a typed client? That is, a client may but is not required to handle defaulting, conversion, or validation?

File: docs/proposals/api-types-package-structure.md [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+The functions apply to only the `group/v1` types, *not* the parent `group/`
+types.
+
+There are two sorts of defaulting functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for `// +default=<value>` comments
+preceding the types or their embedded struct members. Similarly, a
+`// +default=custom` marker will force a stub to be generated. Stubs will be
+generated with a `panic("write me!")` marker to ensure that they are populated.
```
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: See https://github.com/kubernetes/kubernetes/pull/16078/files; defaulting needs a way to say "bail out! don't do any more defaulting on this object!"

File: docs/proposals/api-types-package-structure.md [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+
+```
+apis
+└── group
+    ├── install
+    ├── types.go
+    ├── v1
+    │   ├── install
+    │   │   ├── conversion
+    │   │   │   ├── generated_conversion.go
```
Comment (wojtek-t) [inline docs/proposals/api-types-package-structure.md]: Once we started using ugorji for json encoding we need to put those files somewhere. However, the problem with it is that this tool doesn't support putting this file in another package - the generated file needs to live next the the original file :(

File: docs/proposals/api-types-package-structure.md [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validate_min=<numeric value>,validate_max=<numeric value>`
+* `// +validate_regexp=<regexp>`
+* `// +validate_exact=<exact value>`
```
Comment (wojtek-t) [inline docs/proposals/api-types-package-structure.md]: I assume these are only a subset of directives allowed, right?

Did you think about some more spohisticated things that we are currently validating - e.g. "exactly one of those fields can be set" (volumes) or "two fields need to be equal" (HostPort = ContainerPort)?

Comment (wojtek-t) [inline docs/proposals/api-types-package-structure.md]: Anyway - I think that auto-generating validation functions is a completely different problem, and to be honest I think that we should separate it from the proposal. We should just say that we will be auto-generating something, but the exact mechanism how to do it should be a different proposal. WDYT?

Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: Yeah, this is not a specific proposal, it's more of an example of the sort of thing I want us to do.

I do *not* want the auto-generated-validation to be turing complete, I don't want anything complicated enough that someone has to look up the documentation to understand what it will do. But we should debate this in a narrower document.

File: docs/proposals/api-types-package-structure.md (hunk 2 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: Yeah, that's the idea-- the client can import as much or as little of the tree as needed.

File: docs/proposals/api-types-package-structure.md (hunk 4 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: Yeah, I left encoding generation out of this, but I'd like encoding/decoding generator(s) to follow the same package structure. We can write our own if necessary. I assume we will write our own when we want to do binary encodings.

File: docs/proposals/api-types-package-structure.md (hunk 1 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: OK, I will think about this. Does `apis/<group>/<version>/<resource>/type.go` work for you? Or does it have to be `apis/<group>/<resource>/<version>/type.go`?

Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: One of us is going to have to significantly move things, and it's pretty hard for me to argue that you should combine your stuff into one file, so I guess we'll have to split...

File: docs/proposals/api-types-package-structure.md [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+associated generated files.
+
+Explicit goals are:
+* Organizing our chaos.
+* Make the toolchain easy to understand and use.
+* Reduce the package import graph to a minimum for any given use case.
+* Make it possible to set up a Kubernetes-style types directory in an arbitrary
+  location; decouple our tools and github repository.
+* Coordinate among those who will or might write autogeneration.
+* Where does validation code go? How is it called?
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: I'd like to also address construction

File: docs/proposals/api-types-package-structure.md [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+### Package: `apis/group/`
+
+Multiple versions live in a group. Each version gets its own subdirectory.
+
+Additionally, a types.go file contains an "internal" version of the group's
+objects. This file is used as a destination/source for conversion functions.
+
+### Package: `apis/group/install`
+
+This package installs *all* of the versions of the group.
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: Could clarify what "install" means.

File: docs/proposals/api-types-package-structure.md [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+`types.go` contains the types for this version of the api group.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved to `pkg/apis/unversioned`) and
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: There is no "unversioned" in the tree you drew above.  Is that the internal types or the core stuff like TypeMeta?

File: docs/proposals/api-types-package-structure.md (hunk 2 above)
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: Does "little else" include things like all our util types (IntOrString, Time, etc) ?

File: docs/proposals/api-types-package-structure.md [hunk 9]
```diff
@@ -0,0 +1,381 @@
...
+little else. In particular, it *must not* import its parent `apis/group`.
+
+### Package: `apis/group/v1/install`
+
+This package, if imported, registers the v1 version of this group, its
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: Is it a side-effect of import (`init()`) or an intentional call to a well-known function?  Maybe not a huge deal yet, but it would be nice it this only had intra-group-deps (i.e we it doesn't access global variables in the api pkg (Scheme, ahem)).

File: docs/proposals/api-types-package-structure.md [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/install/conversion`
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: I don't like that this is nested under install - it just doesn't seem obvious.  I also am not sure about the idea that this will register if imported.  Do we have cases where we want to register conversions but NOT defaulting (for example)?

What I was thinking was something like:

`apis/group/v1/install/install.go` imports `apis/group/v1/{conversion,defaults,validation}` and calls `{conversion,defaults,validation}.Register(args)`.  This leaves a minimum of boilerplate gunk inside `{conversion,defaults,validation}/` and accumulates it all in `install/`.  Then we can de-boilerplate that if needed.

If we're going to spend energy making  small well-defined pkgs, let's really try to push the noisy code out of them.

File: docs/proposals/api-types-package-structure.md [hunk 11]
```diff
@@ -0,0 +1,381 @@
...
+
+This package contains conversion functions, and if imported, will register them.
+
+There are two sorts of conversion functions: those which could be automatically
+generated, and those which have manual additions. All the conversion functions
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: Do we have any conversions for which we can't auto-generate SOMETHING?  "could be automatically generated" makes it sound as if we do.

I really like calling this out specifically as "manual overrides".  That is the pattern we should follow - I am certain there are conversions that are getting done wrongly because a new field was added and the conversion was not updated.  every conversion should start with the auto-generated and then make tweaks on that.

File: docs/proposals/api-types-package-structure.md (hunk 3 above)
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: I don't see that in the linked PR?

Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: I'm a little unclear how this will work.  I guess you can emit a function IFF that function doesn't exist, but what happens when types get renamed.  It's a little tricky.  I have a second agenda I want to push - we should not check in generated code.  So we need to distinguish between generated and seeded.

Using comments is cute, but it's going to turn up in godoc - can we do better, such that docs stay user-friendly but the structs can still be the source of truth?

I really want it to be a compile failure if someone specifies +default=custom and then doesn't implement it.  So I think we want to do something like compile-time-generate functions that call the specifically-named default functions.  I'll propose that the specially-named functions might be methods on the type.

We also need to decide on the intended usage of the types in versioned pkgs.  Brian was pushing (and convinced me it was feasible) for no internal types.  Is it desirable for someone to be able to instantiate a foogroup.FooBar type and apply defaults?  Should that be easy?  It starts to be a client-library problem at this point.  Are the versioned client libs going to use these types directly or will they wrap/embed/reimplement them?

File: docs/proposals/api-types-package-structure.md [hunk 12]
```diff
@@ -0,0 +1,381 @@
...
+*only* the object, not cluster-wide or other constraints, which must be validate
+elsewhere. (TODO: copy-pasta this 2x above)
+
+There are two sorts of validation functions, those that can be automatically
+written and those that need human attention.
+
+The former will be autogenerated by looking for directives in the comments
+preceding the types or their embedded struct members. Allowed directives will
+include:
+* `// +validate_min=<numeric value>,validate_max=<numeric value>`
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: same concerns about generated docs and generated vs seeded code.

File: docs/proposals/api-types-package-structure.md (hunk 5 above)
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: Agree with @wojtek-t - we might eventually have declarative validators, but we should start with every field being custom.

We need clarity on what it means to default or validate a _field_ vs a _type_.  For primitives, obviously we are validating each instance of int or string independently.  What about typedef'ed primitives?  If I typedef FooBar as string, should I apply the same validation to every FooBar or does validation differ for every instance of FooBar?   What about structs?  It seems pretty obvious that you don't want to validate different instances of FooStruct differently...

File: docs/proposals/api-types-package-structure.md [hunk 13]
```diff
@@ -0,0 +1,381 @@
...
+* `// +validate_regexp=<regexp>`
+* `// +validate_exact=<exact value>`
+
+If no validation directive is in the comments, a stub with a
+`panic("write me!")` marker will be generated. To prevent this, a
+`// +no_validation_required` directive may be added to the comment.
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: can we just keep a single magic tag, if we go this route?  `validation: none` or somethign

File: docs/proposals/api-types-package-structure.md [hunk 14]
```diff
@@ -0,0 +1,381 @@
...
+    ├── deep
+    ├── default
+    └── validate
+```
+
+These packages will live in the kubernetes repo (or another repo that the
+kubernetes team will own and maintain). The interface for all the packages is
+similar: there'll be a `Register(...)` function for the install packages to use,
+and some function to perform the action. So, e.g., `convert.Convert(&in, &out)`,
+`bar := deep.Copy(&foo)`, `default.Apply(&object)`,
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: all good except the deep copy - do we really want to return interface{} if we can help it///

File: docs/proposals/api-types-package-structure.md [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+of the API group, which may have extra features.
+
+## Type support package layout
+
+The above is how to arrange the types for a particular API. But we also must
+consider, what Kubernetes packages do you need to import to use the API types?
+
+```
+pkg
+└── typetools
```
Comment (thockin) [inline docs/proposals/api-types-package-structure.md]: I feel like this should say about "api" somewhere

File: docs/proposals/api-types-package-structure.md (hunk 1 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: @smarterclayton Further thoughts: if we make groups contain very few resources, would you still want the resources in separate directories?

Comment (smarterclayton) [inline docs/proposals/api-types-package-structure.md]: Our resource directories are groups - image, for instance, is 5 different
API resources working together (ImageStream, Image, ImageStreamTag,
ImageStreamMapping, and ImageStreamImage), and builds are 2 + a bunch of
supporting sub resources.  The only connections between the groups are
object references today (and code in the controllers which knows how to
handle those other groups).  I think it's beneficial to have the entire API
group together in one directory (in the future allows us to run it
independently) but it's not truly required.

Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: I want to keep the implementation separate from the type definition to minimize the size of imports.

@smarterclayton what do you think about having the types for a group/version in the same package but different files? I assume this won't wrok for you, but thought it's worth asking.

File: docs/proposals/api-types-package-structure.md (hunk 6 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: I think I'm gonna say that construction is outside the scope of this, and we should figure that out when we feel like doing away with the "internal" versions of objects.

File: docs/proposals/api-types-package-structure.md (hunk 14 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: `deep.Copy(&foo, &bar)` is great until you don't actually know the concrete type of `foo`, so maybe we should supply both versions?

**Comment by lavalamp** [comment]
@thockin @smarterclayton I did a bit more work on this, PTAL

File: docs/proposals/api-types-package-structure.md (hunk 1 above)
Comment (smarterclayton) [inline docs/proposals/api-types-package-structure.md]: If we argued today that each of the resource types we have today is really an API group, that would just mean we'd have to give up our monolithic root API group.  That's not a bad thing (logically, we expected builds to be distinct from images), although until the infrastructure for groups gets better I wouldn't want to change it aggressively.  Since this is generation, if you didn't support that scenario we could simply live without it until our next iteration of APIs.

Of course, if the *output* is a single directory that's fine for us today.  It's really only the assumption that our types live in a particular package (vs being registered on the scheme) that causes us an issue.

File: docs/proposals/api-types-package-structure.md [hunk 16]
```diff
@@ -0,0 +1,518 @@
...
+  example).
+* Some sort of enabling method, for actually turning a thing on. Registration
+  and enabling are conflated in our codebase at the moment, and this makes it
+  difficult to, for example, construct the default value for --runtime-config.
+* Some sort of execution method, for actually using a thing.
+
+## Reusable apiserver infrastructure packages
+
+In the future, it'd be nice if we could generate REST install code based on
+comment annotations in the go types. See #16560 for a problem statement.
```
Comment (caesarxuchao) [inline docs/proposals/api-types-package-structure.md]: Nit: could you add the link to https://github.com/kubernetes/kubernetes/issues/16560 here? Github doesn't automatically generate the links in .md.

File: docs/proposals/api-types-package-structure.md [hunk 17]
```diff
@@ -0,0 +1,518 @@
...
+
+## Desired toolchain
+
+Along with packages for runtime use, we will supply tools to do the various
+generations.
+
+Ideally, the tool will take the root `apis/` directory, discover the `group`s
+and `version`s, and perform the generations outlined above. We will generate as
+much as possible. The goal is for you to write a `apis/group/types.go` file and
+an `apis/group/v1alpha1` file, push a button, and get the rest of the package
```
Comment (caesarxuchao) [inline docs/proposals/api-types-package-structure.md]: I think you mean "an \`apis/group/v1alpha1/types.go\` file".

**Comment by caesarxuchao** [comment]
Can we make a separate package for Scheme to avoid circular import (original discussion is here: https://github.com/kubernetes/kubernetes/pull/17982/files#r46227185)? Currently it's in pkg/api/register.go (package `api`). It doesn't make sense logically either, because Scheme does not belong to the the internal v1 legacy API. Perhaps we can move it to pkg/apis/scheme?

**Comment by smarterclayton** [comment]
The "registration" approach of the scheme is dangerously close to a
singleton.  It would be better if each API group version exposed a static
Register(*runtime.Scheme) method that was invoked by api/latest or
api/registered that callers used.  I won't force that now - but as soon as
we start dealing with a larger set of extension resources it will become an
issue.

On Dec 1, 2015, at 1:38 AM, Chao Xu <notifications@github.com> wrote:

Can we make a separate package for Scheme to avoid circular import
(original discussion is here:
https://github.com/kubernetes/kubernetes/pull/17982/files#r46227185)?
Currently it's in pkg/api/register.go (package api). It doesn't make sense
logically either, because Scheme does not belong to the the internal v1
legacy API. Perhaps we can move it to pkg/apis/scheme?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16062#issuecomment-160871520>
.

**Comment by lavalamp** [comment]
@thockin Would you like to make another pass?

File: docs/proposals/api-types-package-structure.md [hunk 18]
```diff
@@ -0,0 +1,518 @@
...
+consider, what Kubernetes packages do you need to import to use the API types?
+
+```
+k8s.io/api_machinery/
+└── typetools
+    ├── convert
+    ├── deep
+    ├── default
+    └── validate
+```
```
Comment (caesarxuchao) [inline docs/proposals/api-types-package-structure.md]: Should we add `latest` under `k8s.io/api_machinery/`?

File: docs/proposals/api-types-package-structure.md [hunk 19]
```diff
@@ -0,0 +1,518 @@
...
+
+This package installs *all* of the versions of the group, meaning it imports
+each nested install/ package.
+
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+Multiple resource directories (we show 'resource1' and 'resource2') each have a
+`types.go`, containing the type definitions for this group/version/resource.
```
Comment (liggitt) [inline docs/proposals/api-types-package-structure.md]: a package per resource seems excessive to me... if we achieve our goal of having more granular groups, we shouldn't end up with excessive numbers of resources in a single group, right?

File: docs/proposals/api-types-package-structure.md [hunk 20]
```diff
@@ -0,0 +1,518 @@
...
+resource in this group, its conversion, validation, and defaulting functions.
+
+We have separate install packages for every version to allow deliberate import
+choices to be made.
+
+### Package: `apis/group/v1/resource1/conversion`
+
+This package contains conversion functions, and exports a Register() function
+which will register them (but does not register them as a side effect of being
+imported).
```
Comment (liggitt) [inline docs/proposals/api-types-package-structure.md]: what would it mean for `group/v1/...` to be installed without the corresponding conversion functions being registered? I wouldn't expect to need to register conversions/defaulting after having installed a group/version

Comment (liggitt) [inline docs/proposals/api-types-package-structure.md]: nevermind, it's late and I apparently can't read... install calls the Register() functions

File: docs/proposals/api-types-package-structure.md (hunk 19 above)
Comment (caesarxuchao) [inline docs/proposals/api-types-package-structure.md]: I think we have decided to make a package per GroupVersion. The proposal needs to be updated.

File: docs/proposals/api-types-package-structure.md (hunk 1 above)
Comment (liggitt) [inline docs/proposals/api-types-package-structure.md]: @smarterclayton our packages aren't quite at the resource level... we have several related resources in each package (pretty much what I envision groups being)

File: docs/proposals/api-types-package-structure.md [hunk 21]
```diff
@@ -0,0 +1,518 @@
...
+much as possible. The goal is for you to write a `apis/group/types.go` file and
+an `apis/group/v1alpha1` file, push a button, and get the rest of the package
+structure, including tests that check and enforce the package imports.
+
+We'll deliver a tool with few enough command line arguments that you won't feel
+compelled to wrap it with a bash script. It will be callable via `go generate`.
+
+The tool should never fail to run because *its own* output from a previous
+invocation is missing or broken, but it may fail (with helpful error messages)
+if the input files with user types/functions don't compile.
```
Comment (liggitt) [inline docs/proposals/api-types-package-structure.md]: > The tool should never fail to run because *its own* output from a previous invocation is missing or broken

+1000

File: docs/proposals/api-types-package-structure.md (hunk 19 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: Yes, I misunderstood something Clayton previously said. Should be fixed now.

File: docs/proposals/api-types-package-structure.md (hunk 18 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: I added a paragraph that mentions this below.

File: docs/proposals/api-types-package-structure.md [hunk 22]
```diff
@@ -0,0 +1,473 @@
...
+### Package: `apis/group/v1`
+
+'v1' is taken to be the current stable version of this api.
+
+A `types.go` file containing the v1 version of the group's objects lives here.
+There may be types for several different resources in this `types.go` file.
+
+This package has tightly controlled imports. It's allowed to include the
+Kubernetes unversioned API types (to be moved in/under `pkg/apis/unversioned`)
+and little else. In particular, it *must not* import its parent `apis/group`.
```
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: @deads2k Please take a look at this doc and see what you think :)

File: docs/proposals/api-types-package-structure.md (hunk 1 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: Yup, this is updated now. In case it's not explicit in the doc, you'd be able to have as many /apis/group/... tree structures as you need. Our tools will ideally find them automatically, if not they'll at least let you specify multiple directories.

File: docs/proposals/api-types-package-structure.md (hunk 17 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: yeah, will fix.

File: docs/proposals/api-types-package-structure.md (hunk 16 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: done.

File: docs/proposals/api-types-package-structure.md (hunk 22 above)
Comment (deads2k) [inline docs/proposals/api-types-package-structure.md]: > @deads2k Please take a look at this doc and see what you think :)

Ah, the source of double secret probation.

Comment (deads2k) [inline docs/proposals/api-types-package-structure.md]: I think that one external version should be allowed to import another external version.  For instance, let's say I make a construct that includes a `PodSpec`.  I should be allowed to import the existing `PodSpec` and not have to make a copy.  That has an impact on what the `install` package should do.

**Comment by mattfarina** [comment]
I was looking this over and something struck me.

The document notes a goal of:
> Reduce the package import graph to a minimum for any given use case.
This may be true at program build time but all of kubernetes will need to be downloaded and the entire dependency graph will need to be installed to get at this small subset of packages. For a developer building an application and not working on kubernetes that is a bit of extra work. It becomes more complicated when they have to manage versions of packages in their own applications that are different from those in kubernetes.

Please excuse me if this was discussed somewhere else. The way SDKs, language bindings, and interaction libraries usually work is that there is a versioned API for a service (that's versioned differently from the services), it's documented well, and then there are a variety of SDKs, languages bindings, etc that are designed to talk to it. Each of those are versioned independently for their appropriate cases.

Why is the package Go applications are expected to import deviating from this common model? And, why do develop an app against the kubernetes API do I need to download all of kubernetes and deal with it's dependency tree? What am I missing?

**Comment by lavalamp** [comment]
@mattfarina I don't really follow what you're saying. The goal of this proposal is to get the code into a state where it's possible to grab and use only the parts you need. E.g., if you're writing a client, you don't want server code in your project. Today it's more or less a tangled mess.

> ...  but all of kubernetes will need to be downloaded and the entire dependency graph will need to be installed to get at this small subset of packages.

This is not necessarily true, and fixing it (splitting into separate repos) is another step beyond this proposal.

> The way SDKs, language bindings, and interaction libraries usually work is that there is a versioned API for a service (that's versioned differently from the services), it's documented well, and then there are a variety of SDKs, languages bindings, etc that are designed to talk to it.

This is basically true of kubernetes today, if you don't want to reuse our go client library.

> And, why do those developing an app against the kubernetes API need to download all of kubernetes and deal with it's dependency tree? What am I missing?

Corporations (including google and red hat) want to reuse the go client. Arbitrary people will want to reuse apiserver libraries. People building controllers will want only the client and controller packages. Etc.

File: docs/proposals/api-types-package-structure.md (hunk 22 above)
Comment (lavalamp) [inline docs/proposals/api-types-package-structure.md]: @deads2k I agree that *some* external versions should be able to import *some other* external versions, but not arbitrarily so-- in particular, there must not be cycles. I would like to be even more restrictive, by e.g. making a "common" group and locating there types that other groups are allowed to reuse.

Comment (deads2k) [inline docs/proposals/api-types-package-structure.md]: > @deads2k I agree that some external versions should be able to import some other external versions, but not arbitrarily so-- in particular, there must not be cycles. I would like to be even more restrictive, by e.g. making a "common" group and locating there types that other groups are allowed to reuse.

I think being restrictive of the groups that can be included ends up in a place where you're becoming a gatekeeper on "blessed" types.  I'd like to avoid doing that, but obviously "no cycles" is a rule that makes a lot of sense.

**Comment by eparis** [comment]
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours

File: docs/proposals/api-types-package-structure.md [hunk 23]
```diff
@@ -0,0 +1,475 @@
...
+│           └── manual_validation_overrides.go
+└── unversioned
+    ├── meta
+    │   └── meta.go
+    ├── resource
+    │   └── quantity.go
+    ├── time
+    │   └── time.go
+    └── types.go
+```
```
Comment (krousey) [inline docs/proposals/api-types-package-structure.md]: We now also have encoding and decoding code generation in the same package as types.go. This is only really necessary for the server or high-performance clients. It would be nice to make it optional as well.

Currently, it's implemented as methods on the types. To separate this, you'd have to type aliases in another package, generate the methods there, and have a serializer/deserializer smart enough to do the type conversions.

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!

**Comment by deads2k** [comment]
@lavalamp Did something like this merge?  Relatively few people are familiar with what it takes to add a group.

**Comment by soltysh** [comment]
Or a version inside a group (/me talking about his experiences in adding `v2alpha1` in `batch`)

**Comment by fejta** [comment]
Thanks for the comments. It looks like there is still interest in this PR. @lavalamp will you triage one way or another?

File: docs/proposals/api-types-package-structure.md [hunk 24]
```diff
@@ -0,0 +1,475 @@
...
+of the API group, which may lack features.
+
+### Package: `apis/group/v2alpha1`
+
+This package layout is the same as `apis/group/v1`, but it is an newer version
+of the API group, which may have extra features. It may have types for an
+additional resource, for example.
+
+It is a rule that a client should never have to pay attention to multiple
+versions of a group at the same time, so v2alpha1 copies all types from v1, even
```
Comment (soltysh) [inline docs/proposals/api-types-package-structure.md]: How this applies to your suggestion captured in https://github.com/kubernetes/kubernetes/issues/37166?

**Comment by fejta** [comment]
Can we /close this?
